File: frontend/src/api/projectFlow.ts
import http from './http'
import type { ContractReviewRecordItem } from './contractReviews'

export interface ProjectUpdateLogItem {
  id: number
  operator_user_id: number
  operator_user_name?: string | null
  changed_fields: string
  created_at: string
}

export interface ProjectFlowData {
  project: {
    id: number
    project_no: string
    project_name: string
    client_name: string
    undertaking_unit: string
    evaluation_business_nature?: string | null
    report_type?: string | null
    valuation_base_date?: string | null
    business_salesman?: string | null
    project_amount?: number | null
    invoiced_amount?: number
    project_source: 'INTERNAL' | 'EXTERNAL'
    project_source_display: string
    external_project_leader_name?: string | null
    project_leader_id?: number | null
    project_leader_display_name?: string | null
    display_project_leader_name?: string | null
    contract_no?: string | null
    report_no?: string | null
    first_reviewer_name?: string | null
    second_reviewer_name?: string | null
    third_reviewer_name?: string | null
    print_room_handler_name?: string | null
    invoice_handler_name?: string | null
    archive_reviewer_name?: string | null
    mailing_handler_name?: string | null
    current_step: string
    status_display: string
  }
  current_work_order_id?: number
  current_work_order_status?: string | null
  current_handler_user_id?: number | null
  contract_reviewer_id?: number | null
  contract_reviewer_name?: string | null
  contract_review_status?: string | null
  contract_review_status_display?: string | null
  first_reviewer_id?: number | null
  second_reviewer_id?: number | null
  third_reviewer_id?: number | null
  signer_one?: string | null
  signer_two?: string | null
  formal_report_count?: number | null
  print_room_handler_id?: number | null
  archive_reviewer_id?: number | null
  archive_submitter_id?: number | null
  archive_submission_type?: string | null
  mailing_handler_user_id?: number | null
  mailing_status?: string | null
  signoff_status?: string | null
  chief_appraiser_user_id?: number | null
  user_role_in_project: string
  available_action: string
  can_operate: boolean
  flow_steps: string[]
  contract_review_records: ContractReviewRecordItem[]
  project_update_logs: ProjectUpdateLogItem[]
  review_submit_locked?: boolean
  review_submit_lock_reason?: string | null
  duplicate_delete_required?: boolean
  can_remind_current_handler?: boolean
  reminder_summary?: {
    is_reminded?: boolean
    remind_count_today?: number
    latest_remind_at?: string | null
  } | null
}

export async function getProjectFlow(projectId: number | string) {
  const { data } = await http.get(`/projects/${projectId}/flow`)
  return data as ProjectFlowData
}

File: backend/app/schemas/project_flow.py
from pydantic import BaseModel

from app.schemas.contract_review import ContractReviewRecordResponse


class ProjectFlowProject(BaseModel):
    id: int
    project_no: str
    project_name: str
    client_name: str
    undertaking_unit: str
    evaluation_business_nature: str | None = None
    report_type: str | None = None
    valuation_base_date: str | None = None
    business_salesman: str | None = None
    project_amount: float | None = None
    invoiced_amount: float = 0
    project_source: str = "INTERNAL"
    project_source_display: str = "评估一部"
    external_project_leader_name: str | None = None
    project_leader_display_name: str | None = None
    display_project_leader_name: str | None = None
    contract_no: str | None = None
    report_no: str | None = None
    first_reviewer_name: str | None = None
    second_reviewer_name: str | None = None
    third_reviewer_name: str | None = None
    print_room_handler_name: str | None = None
    mailing_handler_name: str | None = None
    invoice_handler_name: str | None = None
    archive_reviewer_name: str | None = None
    current_step: str
    status_display: str


class ProjectUpdateLogItem(BaseModel):
    id: int
    operator_user_id: int
    operator_user_name: str | None = None
    changed_fields: str
    created_at: str


class ProjectFlowResponse(BaseModel):
    project: ProjectFlowProject
    current_work_order_id: int | None = None
    current_work_order_status: str | None = None
    current_handler_user_id: int | None = None
    contract_reviewer_id: int | None = None
    contract_reviewer_name: str | None = None
    contract_review_status: str | None = None
    contract_review_status_display: str | None = None
    first_reviewer_id: int | None = None
    second_reviewer_id: int | None = None
    third_reviewer_id: int | None = None
    signer_one: str | None = None
    signer_two: str | None = None
    formal_report_count: int | None = None
    print_room_handler_id: int | None = None
    mailing_handler_user_id: int | None = None
    archive_reviewer_id: int | None = None
    archive_submitter_id: int | None = None
    archive_submission_type: str | None = None
    mailing_status: str | None = None
    signoff_status: str | None = None
    chief_appraiser_user_id: int | None = None
    user_role_in_project: str
    available_action: str
    can_operate: bool
    flow_steps: list[str]
    contract_review_records: list[ContractReviewRecordResponse] = []
    project_update_logs: list[ProjectUpdateLogItem] = []
    review_submit_locked: bool = False
    review_submit_lock_reason: str | None = None
    duplicate_delete_required: bool = False
    can_remind_current_handler: bool = False
    reminder_summary: dict[str, str | int | bool | None] | None = None
